### PR TITLE
[FIRRTL] Fix handling of new-style NLAs in ExtractInstances

### PR DESCRIPTION
**Pull request by fabianschuiki**
Adapt `ExtractInstances` such that it can properly handle the new flavor of NLAs, where non-local annotations would carry a symbol to a hierpath that generally only points at their parent module.

The `ExtractInstances` pass had a few assumptions in its implementation about hierpaths pointing at the annotation target op, and each anno having a unique hierpath op associated with it. The implementation can be simplified quite a bit once support for the old-style NLAs can be ripped out.

This commit also adds explicit tests to check that the old *and* new style of NLAs are working in parallel.

A nasty corner case still exists when extraction requires an NLA to be re-routed because the instance has been extracted past the NLA's original root. If the instance is multiply-extracted, one hierpath gets turned into multiple hierpaths, which will require special handling of *everything* with that hierpath. This commit detects the common case which we can handle and ensures that the hierpath is re-rooted in place, and produces a warning about a potential breakage of hierpath references in the corner case.

### CODE REVIEW

File: lib/Dialect/FIRRTL/Transforms/ExtractInstances.cpp [hunk 1]
```diff
@@ -670,26 +683,53 @@ void ExtractInstancesPass::extractInstances() {
...
+            auto newNla = builder.create<HierPathOp>(
+                newInst.getLoc(), circuitNamespace.newName(nla.sym_name()),
+                builder.getArrayAttr(nlaPath));
+            for (auto anno : instNonlocalAnnos.lookup(nla)) {
+              anno.setMember("circt.nonlocal",
+                             FlatSymbolRefAttr::get(newNla.sym_nameAttr()));
+              newInstNonlocalAnnos.push_back(anno);
+            }
+
+            nlaTable.erase(nla);
```
Comment (prithayan) [inline lib/Dialect/FIRRTL/Transforms/ExtractInstances.cpp:715]: Can we remove the `nla` here, is it possible that the `nla` is reused by other annotations ?

**Review by prithayan — APPROVED** [review]
Overall this looks good.

File: test/Dialect/FIRRTL/extract-instances.mlir [hunk 2]
```diff
@@ -65,17 +65,27 @@ firrtl.circuit "ExtractBlackBoxesSimple" attributes {annotations = [{class = "fi
...
-  // CHECK-NOT: firrtl.hierpath @nla_3
-  firrtl.hierpath @nla_1 [@BBWrapper::@bb, @MyBlackBox]
-  firrtl.hierpath @nla_2 [@DUTModule::@mod, @BBWrapper::@bb]
-  firrtl.hierpath @nla_3 [@ExtractBlackBoxesSimple2::@dut, @DUTModule::@mod, @BBWrapper::@bb]
+  // Old style NLAs
+  firrtl.hierpath @nla_old1 [@DUTModule::@mod, @BBWrapper::@bb]
+  firrtl.hierpath @nla_old2 [@ExtractBlackBoxesSimple2::@dut, @DUTModule::@mod, @BBWrapper::@bb]
+  // New style NLAs on extracted instance
+  firrtl.hierpath @nla_on1 [@DUTModule::@mod, @BBWrapper]
+  firrtl.hierpath @nla_on2 [@ExtractBlackBoxesSimple2::@dut, @DUTModule::@mod, @BBWrapper]
```
[the remark this answers is not in the document]
Comment (fabianschuiki) [reply test/Dialect/FIRRTL/extract-instances.mlir:73]: We might want to enforce the new style the same way we do for wires/regs. That would mean a hierpath always ends at the parent module, irrespective of the op, and the last element of a hierpath is alwyas a `SymbolRef`.

Otherwise we again have this degree of freedom where an NLA on an instance *could* mention the instance name in the hierpath, but it wouldn't have to. Forcing one or the other should allow for simpler implementations :thinking:.

Another benefit would be that only hierpaths passing *through* an instance for the sake of addressing something in the subhierarchy actually mentions the instance in the hierpath. Annos that are *on* the instance would have the path terminate earlier, making updating easier.

File: lib/Dialect/FIRRTL/Transforms/ExtractInstances.cpp [hunk 3]
```diff
@@ -670,26 +683,53 @@ void ExtractInstancesPass::extractInstances() {
...
-            newInstNonlocalAnnos.push_back(newAnno);
+          if (instParentNode->hasOneUse()) {
+            // Simply update the existing NLA since our parent is only
+            // instantiated once, and we therefore are not creating multiple
+            // instances through the extraction.
+            nla.namepathAttr(builder.getArrayAttr(nlaPath));
+            for (auto anno : instNonlocalAnnos.lookup(nla))
+              newInstNonlocalAnnos.push_back(anno);
+            nlaTable.erase(nla);
+            nlaTable.addNLA(nla);
```
[the remark this answers is not in the document]
Comment (prithayan) [reply lib/Dialect/FIRRTL/Transforms/ExtractInstances.cpp:700]: Oh, that's right. Another option could be `nlaTable.erase(nla)` can be moved before `nla.namepathAttr(builder.getArrayAttr(nlaPath));`.

File: lib/Dialect/FIRRTL/Transforms/ExtractInstances.cpp [hunk 4]
```diff
@@ -670,26 +683,53 @@ void ExtractInstancesPass::extractInstances() {
...
+                newInst.getLoc(), circuitNamespace.newName(nla.sym_name()),
+                builder.getArrayAttr(nlaPath));
+            for (auto anno : instNonlocalAnnos.lookup(nla)) {
+              anno.setMember("circt.nonlocal",
+                             FlatSymbolRefAttr::get(newNla.sym_nameAttr()));
+              newInstNonlocalAnnos.push_back(anno);
+            }
+
+            nlaTable.erase(nla);
+            nlasToRemove.insert(nla);
```
[the remark this answers is not in the document]
Comment (prithayan) [reply lib/Dialect/FIRRTL/Transforms/ExtractInstances.cpp:716]: `nlaTable.erase()` doesn't erase the op. That has to be done explicitly.

File: lib/Dialect/FIRRTL/Transforms/ExtractInstances.cpp (hunk 1 above)
Comment (fabianschuiki) [reply lib/Dialect/FIRRTL/Transforms/ExtractInstances.cpp:715]: Excellent point! Don't think we can remove it here. Will be fixed in the rebase.

File: lib/Dialect/FIRRTL/Transforms/ExtractInstances.cpp (hunk 3 above)
[the remark this answers is not in the document]
Comment (fabianschuiki) [reply lib/Dialect/FIRRTL/Transforms/ExtractInstances.cpp:700]: Yeah good point. I've moved the erase call up before mutating the NLA as @prithayan suggested :+1:

File: lib/Dialect/FIRRTL/Transforms/ExtractInstances.cpp [hunk 5]
```diff
@@ -670,26 +683,53 @@ void ExtractInstancesPass::extractInstances() {
...
+            nla.namepathAttr(builder.getArrayAttr(nlaPath));
+            for (auto anno : instNonlocalAnnos.lookup(nla))
+              newInstNonlocalAnnos.push_back(anno);
+            nlaTable.erase(nla);
+            nlaTable.addNLA(nla);
+            LLVM_DEBUG(llvm::dbgs() << "    - Modified to " << nla << "\n");
+          } else {
+            // Since we are extracting to multiple parent locations, create a
+            // new NLA for each instantiation site.
+            auto builder = OpBuilder::atBlockBegin(getOperation().getBody());
```
[the remark this answers is not in the document]
Comment (fabianschuiki) [reply lib/Dialect/FIRRTL/Transforms/ExtractInstances.cpp:705]: This makes total sense. No idea why I didn't do it like this in the first place :grimacing:.

File: lib/Dialect/FIRRTL/Transforms/ExtractInstances.cpp [hunk 6]
```diff
@@ -670,26 +683,53 @@ void ExtractInstancesPass::extractInstances() {
...
+            auto newNla = builder.create<HierPathOp>(
+                newInst.getLoc(), circuitNamespace.newName(nla.sym_name()),
+                builder.getArrayAttr(nlaPath));
```
[the remark this answers is not in the document]
Comment (fabianschuiki) [reply lib/Dialect/FIRRTL/Transforms/ExtractInstances.cpp:708]: Excellent idea! Done.

File: lib/Dialect/FIRRTL/Transforms/ExtractInstances.cpp (hunk 4 above)
[the remark this answers is not in the document]
Comment (fabianschuiki) [reply lib/Dialect/FIRRTL/Transforms/ExtractInstances.cpp:716]: I dream of an `NLATable` that insert, delete, and refcount `HierPathOp`s implicitly with no pass ever having to `builder.create<HierPathOp>` ever again :grimacing: :wink:
